ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -175,6 +198,167 @@ def _map_sample(self, sample, exclude_fields):
+    It uses the Montreal Forced Aligner (MFA) to align the text with the audio. The result is a
+    list of timestamps for each word in the text transcription. The timestamps are stored in a new
+    column, in a list of dict format:
+        [ {"start": float in seconds, "end": float in seconds, "text": first word str}, ... ]
+    """
+
+    # Jinja template for the text transcription that needs to be aligned
+    template: str = simple_parsing.field(alias="-T")
+    # The accoustic model to use for MFA alignment.
+    # Make sure the dictionary and acoustic model are installed. See just install_mfa for an example (English).
+    # Model index: https://mfa-models.readthedocs.io/en/latest/acoustic/index.html
+    # For many languages there exists a {language}_mfa model that you can use, e.g. "english_mfa"
+    mfa_acoustic_model: str = simple_parsing.field(alias="-m")
+    # The dictionary to use for MFA alignment. Defaults to the same name as the acoustic model.
+    mfa_dictionary: Optional[str] = simple_parsing.field(default=None, alias="-d")
+    audio_column_name: str = simple_parsing.field(default="audio", alias="-a")
+    sample_rate: int = simple_parsing.field(default=16000, alias="-r")
+    # The column name to store the timestamps in
+    timestamp_column_name: str = simple_parsing.field(default="timestamps", alias="-ts")
+    aligned_ratio_check: float = simple_parsing.field(default=0.95, alias="-ar")
+
+    def __post_init__(self):
+        if self.mfa_dictionary is None:
+            self.mfa_dictionary = self.mfa_acoustic_model
+
+        try:
+            # Make sure the MFA environment is installed
+            subprocess.run(["conda", "run", "-n", MFA_ENV_NAME, "echo"], check=True)
+        except subprocess.CalledProcessError:
+            raise Exception(
+                "Please install the MFA environment using `just install_mfa` first."
+            )
+
+        if self.template.startswith("@"):
+            with open(self.template[1:], "r") as template_file:
+                self.template = template_file.read()
+
+    def map_split(
+        self,
+        ds_split: datasets.Dataset,
+        num_proc: int,
+        writer_batch_size: int,
+        exclude_fields: List[str],
+    ) -> datasets.Dataset:
+        # 0. create a temp directory to store the audio and text files
+        _temp_dir = tempfile.TemporaryDirectory()
+        self.temp_dir = _temp_dir.name
+        os.makedirs(self.temp_dir, exist_ok=True)
+
+        # 1. copy all audio-text pairs into the temp directory
+        ds_split.map(
+            self._store_sample_as_files,
+            num_proc=num_proc,
+            fn_kwargs={"exclude_fields": set(exclude_fields)},
+        )
+
+        count_wavs = len(glob.glob(os.path.join(self.temp_dir, "*.wav")))
+        assert count_wavs >= len(
+            ds_split
+        ), "Not all samples were stored as files. The id is likely not unique."
+
```

Comment:
what is the legitimate case when `count_waves > len(ds_split)`?

Reply:
I think that can only happen if we reuse the temp directory (which I was doing before). I can tighten the condition.

Reply:
Done.